Update previous succeeded indexes for Indexed jobs unconditionally

danielvegamyhre opened the pull request:
#### What type of PR is this?
/kind bug
/priority backlog
/sig apps

#### What this PR does / why we need it:

This PR fixes a bug which caused the status of Indexed Jobs to only be updated when there are newly completed indexes. The completed indexes are now updated if the .status.completedIndexes has values outside of the [0, .spec.completions> range

#### Which issue(s) this PR fixes:
Bug was discovered while implementing #114862

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
Fixed bug which caused the status of Indexed Jobs to only be updated when there are newly completed indexes. The completed indexes are now updated if the .status.completedIndexes has values outside of the [0, .spec.completions> range
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
```

danielvegamyhre committed:
update prev succeeded indexes for indexed jobs unconditionally

ahg-g commented:
/kind bug

ahg-g commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -1502,6 +1580,8 @@ func TestTrackJobStatusAndRemoveFinalizers(t *testing.T) {
 	}
 	for name, tc := range cases {
 		t.Run(name, func(t *testing.T) {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.ScalableIndexedJob, true)()
```
remove this

danielvegamyhre replied:
woops, i meant to only include the feature gate in the PR with the validation change. it is now removed.

mimowo commented on pkg/controller/job/job_controller.go:
```diff
@@ -1696,6 +1696,10 @@ func findConditionByType(list []batch.JobCondition, cType batch.JobConditionType
 func recordJobPodFinished(job *batch.Job, oldCounters batch.JobStatus) {
 	completionMode := completionModeStr(job)
 	diff := job.Status.Succeeded - oldCounters.Succeeded
+	// prometheus CounterVec metric must be monotonically increasing
```
nit: I would suggest to expand the comment here to say that the counter for succeeded pods may decrease when an indexed job is scaled down.

danielvegamyhre replied:
The counter actually won't be decreased when the indexed job is scaled down, since the `diff` would be negative and thus set to 0 on lines 1700-1702. The prometheus counter metric will actually panic if you try to decrease it, which is why I update it by max(0, diff).

alculquicondor replied:
Let's see this scenario:
1. completions = 6, and newSucceededPods={4}
2. metric=1
3. Then, user changes completions=3 and newSucceededPods={2}

The metric should be 2, not 1, because {2} is a newly succeeded pod. Can you fix that?

danielvegamyhre replied:
Ok I figured out a way to address this scenario and updated the code accordingly, let me know what you think.

mimowo reviewed (commented):
LGTM, one optional nit to expand the comment when recording metrics

alculquicondor commented:
/ok-to-test

alculquicondor commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -675,6 +676,25 @@ func TestControllerSyncJob(t *testing.T) {
 			expectedFailed:        0,
 			expectedPodPatches:    5,
 		},
+		"indexed job with pod indexes higher than parallelism has higher indexes removed first": {
```
```suggestion
		"indexed job with pod indexes higher than completions has higher indexes removed first": {
```

alculquicondor replied:
but this is already the test case above "indexed job with indexes outside of range"

danielvegamyhre replied:
oh this unit test was a remnant of the original "allow spec.Completions to be nil" proposal which used parallelism to scale the job while completions is nil. I have removed this test now.

alculquicondor commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -904,14 +924,16 @@ func TestControllerSyncJob(t *testing.T) {
+func checkIndexedJobPods(t *testing.T, control *controller.FakePodControl, wantIndexes, doNotWantIndexes sets.Int, jobName string) {
 	t.Helper()
 	gotIndexes := sets.NewInt()
 	for _, p := range control.Templates {
```
`control.Templates` only has objects that the controller created. It has no information about deleted pods, so the new code is not adding any new coverage.

I would leave it out. Otherwise, you can validate `fakePodControl.DeletePodName`

danielvegamyhre replied:
Ah ok, good to know. Done.

alculquicondor commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -1340,6 +1362,62 @@ func TestTrackJobStatusAndRemoveFinalizers(t *testing.T) {
+			},
+			needsFlush: true,
+		},
+		"indexed job prev successful pods outside current completions index range with new succeeded pods in range": {
```
this case was passing before fixing the controller, right?

danielvegamyhre replied:
yes that's correct

alculquicondor commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -1536,6 +1614,10 @@ func TestTrackJobStatusAndRemoveFinalizers(t *testing.T) {
 					t.Fatalf("Obtaining succeeded job_pods_finished_total: %v", err)
 				}
 				newSucceeded := job.Status.Succeeded - tc.job.Status.Succeeded
```
let's be more explicit and have a new variable `tc.wantSucceededPodsMetric`.

Same for failed.

danielvegamyhre replied:
refactored test to have `tc.wantSucceededPodsMetric` and `tc.wantFailedPodsMetric`

alculquicondor commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -1340,6 +1362,62 @@ func TestTrackJobStatusAndRemoveFinalizers(t *testing.T) {
+					UncountedTerminatedPods: &batch.UncountedTerminatedPods{},
+				},
+			},
+			needsFlush: true,
```
wait, this test should be written with `needsFlush: false`. This will reveal a bug in this PR's implementation

`succeededIndexes = succeededIndexes.withOrderedIndexes(newSucceededIndexes)`

at this point we should set needsFlush=true if there is a change.

danielvegamyhre replied:
Done

alculquicondor commented on pkg/controller/job/indexed_job_utils.go:
```diff
@@ -341,3 +341,47 @@ func completionModeStr(job *batch.Job) string {
+// parseCompletedIndexes parses the interval string from
+// job.Status.CompletedIndexes and returns a map of unique
+// completed indexes.
+func parseCompletedIndexes(completedIndexes string) map[int]bool {
```
can't you use `succeededIndexesFromJob`? Or at least refactor that function to fit the new needs.

Note that using `map[int]` is very problematic, because the number of completions can be very large. `orderedIntervals` is a compressed version of a map.

danielvegamyhre replied:
Makes sense, I have implemented the refactor of `succeededIndexesFromJob ` and updated my approach accordingly.

alculquicondor commented on pkg/controller/job/job_controller.go:
```diff
@@ -1089,13 +1093,13 @@ func (jm *Controller) trackJobStatusAndRemoveFinalizers(ctx context.Context, job
 // Returns whether there are pending changes in the Job status that need to be
 // flushed in subsequent calls.
-func (jm *Controller) flushUncountedAndRemoveFinalizers(ctx context.Context, job *batch.Job, podsToRemoveFinalizer []*v1.Pod, uidsWithFinalizer sets.String, oldCounters *batch.JobStatus, podFailureCountByPolicyAction map[string]int, needsFlush bool) (*batch.Job, bool, error) {
+func (jm *Controller) flushUncountedAndRemoveFinalizers(ctx context.Context, job *batch.Job, podsToRemoveFinalizer []*v1.Pod, uidsWithFinalizer sets.String, oldCounters *batch.JobStatus, podFailureCountByPolicyAction map[string]int, needsFlush bool, prevSucceededIndexes orderedIntervals, newSucceededIndexes []int) (*batch.Job, bool, error) {
```
it looks like you are not using `prevSucceededIndexes`

danielvegamyhre replied:
Woops yes this was a remnant of a prior approach that I tried - I've now removed it.

alculquicondor commented on pkg/controller/job/job_controller.go:
```diff
@@ -1693,11 +1697,32 @@ func findConditionByType(list []batch.JobCondition, cType batch.JobConditionType
+	// now out of range (i.e. index >= spec.Completions).
+	if isIndexedJob(job) {
+		prevUniqueIndexes := parseCompletedIndexes(oldCounters.CompletedIndexes)
+		newUniqueIndexes := uniqueIndexes(newSucceededIndexes)
```
Because of this line:

```golang
if ix != unknownCompletionIndex && ix < int(*job.Spec.Completions) && !succeededIndexes.has(ix) {
```

`newSucceededIndexes` should be a disjoint set from `prevUniqueIndexes`. Then all that matter is how many unique indexes there are in `newSucceededIndexes`.

alculquicondor commented on pkg/controller/job/job_controller.go:
```diff
@@ -1041,8 +1042,11 @@ func (jm *Controller) trackJobStatusAndRemoveFinalizers(ctx context.Context, job
 	}
-	if len(newSucceededIndexes) > 0 {
+	if isIndexed {
 		succeededIndexes = succeededIndexes.withOrderedIndexes(newSucceededIndexes)
```
I think, at this point `succeededIndexes-prevSucceededIndexes` should have the delta.

alculquicondor replied:
Did you try this? `succeededIndexes.total() - prevSucceededIndexes.total()`

danielvegamyhre replied:
Yes, it results in the test case indexed job past and new pods failing due to `recordJobPodFinished()` being called twice during the test, so simply passing in this delta results in double counting.

danielvegamyhre commented:
@alculquicondor I have refactored `succeededIndexesFromJob` to meet the requirements needed here and updated my approach in this PR accordingly, let me know if you have any thoughts

alculquicondor commented on pkg/controller/job/job_controller.go:
```diff
@@ -968,8 +968,8 @@ func (jm *Controller) trackJobStatusAndRemoveFinalizers(ctx context.Context, job
-	// Shallow copy, as it will only be used to detect changes in the counters.
-	oldCounters := job.Status
+
+	oldJob := job.DeepCopy()
```
`DeepCopy` is generally expensive. Why is a shallow copy not enough?

danielvegamyhre replied:
A shallow copy results in most of the tests failing due to this line I believe: https://github.com/kubernetes/kubernetes/pull/115349/files#diff-a7577a7dede7864ff38c631319033714fdd0bed91108d976282e1099507e6ff0R1103 With a shallow copy of `job` the `oldJob.Status` pointer is still pointing to the same address in memory as the original `job` so it is modifying the wrong status.

danielvegamyhre replied:
Refactored succeededIndexesFromJob to accept a `batch.JobStatus` and `completions` and calculate the succeeded metric diff for Indexed Jobs using the *new job spec's completions* like so, which seems to work:

```go
succeededIndexesFromJob(job.Status.CompletedIndexes, int(*job.Spec.Completions)).total() - succeededIndexesFromJob(oldCounters.CompletedIndexes, int(*job.Spec.Completions)).total()
```

alculquicondor commented on pkg/controller/job/job_controller.go:
```diff
@@ -1693,11 +1696,32 @@ func findConditionByType(list []batch.JobCondition, cType batch.JobConditionType
+	// in tandem, and now a previously completed index is
+	// now out of range (i.e. index >= spec.Completions).
+	if isIndexedJob(job) {
+		prevUniqueIndexes := succeededIndexesFromJob(oldJob, false)
```
I'm not convinced that passing `(oldJob, true)` here makes any difference, because `newSucceededIndexes` doesn't include any indexes past `completions`.

danielvegamyhre replied:
> Did you try this? succeededIndexes.total() - prevSucceededIndexes.total()

Yes, it results in the test case `indexed job past and new pods` failing due to `recordJobPodFinished()` being called twice during the test, so simply passing in this delta results in double counting.

danielvegamyhre replied:
> I'm not convinced that passing (oldJob, true) here makes any difference, because newSucceededIndexes doesn't include any indexes past completions.

Looks like you're right! I have updated the code accordingly so `succeededIndexesFromJob` is in its original state, as there was nowhere that I needed to set `filterOutOfRangeIndexes` to false.

alculquicondor reviewed (commented):
Giving a few alternatives to avoid passing extra parameters and avoid a deep copy.

danielvegamyhre commented on pkg/controller/job/job_controller.go:
```diff
@@ -1693,11 +1696,32 @@ func findConditionByType(list []batch.JobCondition, cType batch.JobConditionType
+	// now out of range (i.e. index >= spec.Completions).
+
+	if isIndexedJob(job) {
+		prevCompletedIndexes := succeededIndexesFromJob(oldJob)
+		newSucceeded := 0
+		for idx, _ := range uniqueIndexes(newSucceededIndexes) {
+			if !prevCompletedIndexes.has(idx) {
+				newSucceeded += 1
+			}
+		}
+		diff = newSucceeded
```
Hmm orderedIntervals doesn't have a `len()` method but if I use Go's native 'len` function (i.e. len(succeededIndexesFromJob(...))` I believe it will return the number of intervals, not the number of completed indexes in those intervals - I don't think that is the desired behavior here, right? 

I think orderedIntervals's `.total()` method returns the number of completed indexes in the intervals, and I tried `succeededIndexes.total() - prevSucceededIndexes.total()` already which has issues described in another comment above.

danielvegamyhre replied:
Ok I managed to avoid making the deep copy by doing some light refactoring to simply use the shallow copy of job.Status and the number of completions.

alculquicondor replied:
oops, I meant total :)
I'll take another look

danielvegamyhre commented on pkg/controller/job/job_controller.go:
```diff
@@ -1073,7 +1076,7 @@ func (jm *Controller) trackJobStatusAndRemoveFinalizers(ctx context.Context, job
 			jm.recordJobFinished(job, finishedCond)
 		}
-		recordJobPodFinished(job, oldCounters)
+		recordJobPodFinished(job, oldJob, newSucceededIndexes)
```
Sure, I tried removing this line results in many of the TestTrackJobStatusAndRemoveFinalizers tests failing, so it seems necessary.

danielvegamyhre commented on pkg/controller/job/indexed_job_utils.go:
```diff
@@ -341,3 +341,11 @@ func completionModeStr(job *batch.Job) string {
 	return string(batch.NonIndexedCompletion)
 }
+
+func uniqueIndexes(indexes []int) map[int]bool {
```
done

danielvegamyhre commented on pkg/controller/job/indexed_job_utils.go:
```diff
@@ -148,20 +148,19 @@ func (oi orderedIntervals) has(ix int) bool {
 		var err error
 		inter.First, err = strconv.Atoi(limitsStr[0])
 		if err != nil {
-			klog.InfoS("Corrupted completed indexes interval, ignoring", "job", klog.KObj(job), "interval", intervalStr, "err", err)
```
@alculquicondor will we still want to log the job object here? I can add it back, but I removed the job as a param to the function since it is used for anything besides these log lines and it would make the function signature confusing IMO since we are passing in what appears to be redundant information (i.e. completedIndexes, completions). However if logging the job object in these lines is import I will add it back.

alculquicondor commented on pkg/controller/job/indexed_job_utils.go:
```diff
@@ -148,20 +148,19 @@ func (oi orderedIntervals) has(ix int) bool {
 
-func succeededIndexesFromJob(job *batch.Job) orderedIntervals {
-	if job.Status.CompletedIndexes == "" {
+func succeededIndexesFromJob(completedIndexes string, completions int) orderedIntervals {
```
```suggestion
func succeededIndexesFromString(completedIndexes string, completions int) orderedIntervals {
```

or just `succeededIndexes`, but this could be a bit ambiguous?

danielvegamyhre replied:
I opted for `succeededIndexesFromString` since `succeededIndexes` is used in a variable name in various places already, plus the function name is a bit more specific as you mentioned.

alculquicondor commented on pkg/controller/job/job_controller.go:
```diff
@@ -1041,8 +1042,11 @@ func (jm *Controller) trackJobStatusAndRemoveFinalizers(ctx context.Context, job
+			needsFlush = true
+		}
 		job.Status.Succeeded = int32(succeededIndexes.total())
 		job.Status.CompletedIndexes = succeededIndexes.String()
```
we are calling `String()` twice, which is not very expensive, but not trivial and can be avoided. Save it in variable above.

danielvegamyhre replied:
Done

alculquicondor commented on pkg/controller/job/job_controller.go:
```diff
@@ -1695,9 +1699,22 @@ func findConditionByType(list []batch.JobCondition, cType batch.JobConditionType
+	// in tandem, and now a previously completed index is
+	// now out of range (i.e. index >= spec.Completions).
+	if isIndexedJob(job) {
+		diff = succeededIndexesFromJob(job.Status.CompletedIndexes, int(*job.Spec.Completions)).total() - succeededIndexesFromJob(oldCounters.CompletedIndexes, int(*job.Spec.Completions)).total()
```
We could save the cost of parsing if CompletedIndexes didn't change.

danielvegamyhre replied:
Done

alculquicondor commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -1517,7 +1594,11 @@ func TestTrackJobStatusAndRemoveFinalizers(t *testing.T) {
 			}
 			uncounted := newUncountedTerminatedPods(*job.Status.UncountedTerminatedPods)
-			succeededIndexes := succeededIndexesFromJob(job)
+			completions := 0
+			if isIndexedJob(job) {
+				completions = int(*job.Spec.Completions)
+			}
+			succeededIndexes := succeededIndexesFromJob(job.Status.CompletedIndexes, completions)
```
```suggestion
            var succeededIndexes orderedIndexes
			if isIndexedJob(job) {
				succeededIndexes = succeededIndexesFromJob(job.Status.CompletedIndexes, int(*job.Spec.Completions))
			}
```

danielvegamyhre replied:
Done

alculquicondor commented on pkg/controller/job/job_controller.go:
```diff
@@ -1695,9 +1700,22 @@ func findConditionByType(list []batch.JobCondition, cType batch.JobConditionType
+	// now out of range (i.e. index >= spec.Completions).
+	if isIndexedJob(job) && job.Status.CompletedIndexes != oldCounters.CompletedIndexes {
+		diff = succeededIndexesFromString(job.Status.CompletedIndexes, int(*job.Spec.Completions)).total() - succeededIndexesFromString(oldCounters.CompletedIndexes, int(*job.Spec.Completions)).total()
+	} else {
```
this else has changed meaning now.

danielvegamyhre replied:
Updated

danielvegamyhre commented:
/retest

alculquicondor commented on pkg/controller/job/job_controller.go:
```diff
@@ -1695,9 +1700,26 @@ func findConditionByType(list []batch.JobCondition, cType batch.JobConditionType
+	// now out of range (i.e. index >= spec.Completions).
+	if isIndexedJob(job) {
+		if job.Status.CompletedIndexes == oldCounters.CompletedIndexes {
+			diff = 0
```
it's already zero when declared in line 1703, so you can remove this conditional branch.

danielvegamyhre replied:
Makes sense, updated

alculquicondor commented:
/lgtm
/approve

Please prepare cherry-picks for 1.24+

danielvegamyhre commented:
/retest

alculquicondor commented:
Please open issues for the flaky tests, if there aren't any already.
